File: quasar/src/main/java/ru/relex/techtalks/async/quasar/controller/PatientController.java
package ru.relex.techtalks.async.quasar.controller;

import co.paralleluniverse.fibers.Fiber;
import co.paralleluniverse.fibers.SuspendExecution;
import co.paralleluniverse.fibers.Suspendable;
import co.paralleluniverse.strands.SuspendableCallable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import ru.relex.techtalks.async.model.PatientDto;
import ru.relex.techtalks.async.quasar.service.PatientService;

import java.util.concurrent.ExecutionException;

@RequestMapping("/api")
@RestController
public class PatientController {
    private static final Logger logger = LoggerFactory.getLogger(PatientController.class);

    private final PatientService patientService;

    @Autowired
    public PatientController(PatientService documentService) {
        this.patientService = documentService;
    }

    @Suspendable
    @GetMapping("/ping")
    public String ping() throws Exception {
        return "ok";
    }

    @GetMapping("/patient/{id}/data")
    @Suspendable
    public PatientDto getPatientData(
            @PathVariable("id") Long id,
            @RequestParam("category") String category,
            @RequestParam("location") String location
    ) throws ExecutionException, InterruptedException {
        try {
            return patientService.retrievePatientData(id, category, location);
        } catch (Exception e) {
            logger.error("Error during patient data retrieving", e);
            throw e;
        }
    }
}

File: kotlin/src/main/kotlin/ru/relex/techtalks/async/kotlin/controller/PatientController.kt
package ru.relex.techtalks.async.kotlin.controller

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import ru.relex.techtalks.async.kotlin.service.PatientService
import ru.relex.techtalks.async.model.PatientDto

@RestController
@RequestMapping("/api")
class PatientController @Autowired() constructor(
  private val patientService: PatientService
) {

  private val logger = LoggerFactory.getLogger(PatientController::class.java)

  @GetMapping("/ping")
  suspend fun ping(): String {
    return "ok"
  }

  @GetMapping("/patient/{id}/data")
  suspend fun getPatientData(
    @PathVariable("id") id: Long,
    @RequestParam("category") category: String,
    @RequestParam("location") location: String
  ): PatientDto {
    try {
      return patientService.retrievePatientDataAsync(id, category, location)
    } catch (e: Exception) {
      logger.error("Error during patient data retrieving", e)
      throw e
    }
  }
}
